src/Swarm/Game/Step.hs

```diff
@@ -704,24 +699,22 @@ execConst c vs k = do
 
         not (null recipes) `holdsOr` cmdExn Drill ["There is no way to drill a", nextE ^. entityName <> "."]
 
-        -- add and remove drill as it is requiered in recipe - the drilled entity is replaced at the end
-        let inv' = insert drill $ insert nextE inv
-        (invTaken, outs) <-
-          listToMaybe (rights (map (make' inv') recipes))
+        -- add the drilled entity so it can be consumed by recipe
+        let makeRecipe r = (\(i, o) -> (i, o, r)) <$> make' (insert nextE inv, ins) r
+        (invTaken, outs, recipe) <-
+          listToMaybe (rights (map makeRecipe recipes))
             `isJustOr` cmdExn Drill ["You don't have the ingredients\n to drill", indefinite (nextE ^. entityName) <> "."]
 
         let (out, down) = L.partition ((`hasProperty` Portable) . snd) outs
-            downAmount = sum $ map fst down
 
-        (downAmount <= 1) `holdsOr` cmdExn Drill ["Bad recipe - can not place more then one unmovable entity."]
+        case down of
+          [] -> updateEntityAt nextLoc (const Nothing)
+          [de] -> updateEntityAt nextLoc (const (Just $ snd de))
+          _ -> throwError $ cmdExn Drill ["Bad recipe - can not place more then one unmovable entity."]
 
-        robotInventory .= delete drill (L.foldl' (flip (uncurry insertCount)) invTaken out)
-        _ <-
-          if downAmount == 0
-            then updateEntityAt nextLoc (const Nothing)
-            else updateEntityAt nextLoc (const (Just . snd . head $ down))
+        robotInventory .= L.foldl' (flip (uncurry insertCount)) invTaken out
         flagRedraw
-        return $ Out VUnit k
+        finishCookingRecipe recipe
```

Comment:
Recipes can now take time (using `Waiting`), though they do change the world first.
Not sure if that weird effect can be avoided, but at least it can not be cancelled with `Ctrl+c` :smile:

Reply:
I see you noticed too @byorgey :smile: maybe it is not a feature after all :bug: 
- #194

Reply:
Oh! I noticed independently.  No, I do not think it is a feature.  I don't see why I shouldn't be allowed to change my mind about waiting. =)

Reply:
I definitely think all the effects need to happen *after* the waiting period finishes, otherwise it seems weird (if the recipe already worked right away, why do we still need to wait?) and open to glitches that circumvent the waiting period.

I think the way to do this is to add a new kind of continuation stack frame.  The continuation stack says "what to do next" once you are done evaluating the current thing; in this case we want to `wait` while remembering that after done waiting, we need to make some changes to the game state.
* Create a new constructor for the `Frame` type in `Swarm.Game.CEK`, e.g. `FFinishRecipe`.  Shove whatever information is needed into it to actually carry out the effects of a recipe (e.g. I guess we need some things to take away from the robot's inventory and some things to add, and maybe also something about how to update the world?).
* Add a case to the `stepCEK` function: if we see `Out v (FFinishRecipe blah blah : k)`, then do the stuff (insert and delete from the inventory, modify the world, etc.) and finally return `Out v k`.
* Now in `finishCookingRecipe`, return something like `Out VUnit (FFinishRecipe blah blah : k)` instead of `Out VUnit k`.